### PR TITLE
Dev run hooks

### PR DESCRIPTION
**Pull request by iilyak**
**NOT** ready for merge.

Opening for discussion **only**.

- Somewhat related to https://github.com/apache/couchdb/issues/1515

Every developer might have a slightly different setup. Currently in order to do this customization developers are forced to modify `dev/run` and some other files. This PR is proposing to introduce a `hooks_dir` for `dev/run` customizations. The hooks are simple python modules which export following set of functions:

- `before_setup(ctx)`
- `after_setup(ctx)`
- `before_boot_nodes(ctx)`
- `after_boot_nodes(ctx)`
- `before_startup(ctx)`
- `after_startup(ctx)`

### CODE REVIEW

File: dev/run [hunk 1]
```diff
@@ -88,10 +94,13 @@ def main():
 def setup():
     opts, args = setup_argparse()
     ctx = setup_context(opts, args)
+    print ctx
+    apply_hooks_from_dir(ctx['hooks_dir'], 'before_setup', ctx=ctx)
```
Comment (iilyak) [inline dev/run]: we might want to be able to allow updating of ctx by the hook.

**Comment by nickva** [comment]
I like the idea. Might be useful in setting up some manual scenarios or test scripts. Say fill the cluster nodes with data in after_boot_nodes.

**Comment by wohali** [comment]
This is the first I'm hearing that people have to modify `dev/run`. Can you provide some context? Is this a Cloudant-specific thing?

**Comment by wohali** [comment]
Also, if you're going to run custom hooks in `hooks_dir`, please use `.gitignore` so that stuff doesn't get accidentally commited. Ideally the entire `hooks_dir` would be commited empty, because people are likely to mod your sample `dev/hooks/01_echo.py` file and accidentally commit that, too.

**Comment by eiri** [comment]
I'm -0 on this. I don't have use for such hooks, the only time I needed a custom setup was during db data format migration when I ran two different versions of Couch over the same data dir and I found it reasonable that I had to build a custom setup instead of modifying `dev/run`. And these days I just run docker-compose setup when I need a custom local cluster.

**Comment by iilyak** [comment]
I am going to close it since it is not extremely important feature and complicates `dev/run`

**Pull request closed without merging**